File: 133. Clone Graph/Java/Solution.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        class Utils {
            UndirectedGraphNode clone(UndirectedGraphNode node, Map<Integer, UndirectedGraphNode> memory) {
                if (!memory.containsKey(node.label)) {
                    UndirectedGraphNode copy = new UndirectedGraphNode(node.label);
                    memory.put(copy.label, copy);
                    for (UndirectedGraphNode neighbor : node.neighbors) {
                        copy.neighbors.add(clone(neighbor, memory));
                    }
                }
                return memory.get(node.label);
            }
        }
        if (node == null) {
            return node;
        }
        Utils utils = new Utils();
        Map<Integer, UndirectedGraphNode> memory = new HashMap<>();
        return utils.clone(node, memory);
    }
}

File: 133. Clone Graph/Python/Solution.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []


class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        def clone(node, memory):
            if node.label not in memory:
                copy = UndirectedGraphNode(node.label)
                memory[copy.label] = copy
                for neighbor in node.neighbors:
                    copy.neighbors.append(clone(neighbor, memory))
            return memory[node.label]
        if not node:
            return node
        memory = {}
        return clone(node, memory)
